### ISSUE TITLE
make it async

### ISSUE THREAD

**Issue by glyph** [asottile/babi#79]
The ability to asynchronously populate the contents of a buffer from a subprocess, and to update things without blocking, is a  key feature of an extensible editor with a good user experience.  Otherwise, extensions have a tendency to block the user interface and generally bog down as you add more of them.

Please consider adding support for this.  Thanks!

(I dream of one day replacing my reams and reams of emacs lisp with Python, and Babi seems like a promising contender for that ;-))

**Comment by asottile** [comment]
makes sense!  I think I was getting pretty close to hitting this (adding some support for linters / plugins is pretty high on the list for "next things")

I *might* have to investigate writing some sort of async-friendly wrapper around curses since I'm not aware of anything that exists in that space 🤔

**Comment by glyph** [comment]
`twisted.python.conch.insults` provides an async terminal windowing library, which would also let you SSH into the editor to use it from multiple sessions at once 😎 .

I don't know if I'd seriously recommend it though unless you wanted to help maintain such a thing; it's unfortunately bereft of maintainers and somewhat thinly documented.

More practical is probably just sticking the `curses` file descriptor into an asyncio event loop, and using `nodelay`: https://docs.python.org/3/howto/curses.html#user-input

**Comment by asottile** [comment]
yeah that makes sense, there's already some use of `nodelay` in babi for reading some specialized key sequences and it would probably be ~reasonably easy to adapt the central loop for that

I also want (at some point) to move the first-time syntax highlighting procedure to a background-ish task which would fix the current (major) lag (open large file, jump to end)

**Comment by graingert** [comment]
urwid is pretty nice and supports trio, asyncio and twisted etc
